File: tutorials/003_benchmark_your_model/benchmark_your_model.cpp
// Benchmark a compiled model with deterministic synthetic inputs.
//
// Usage:
//   tutorial_003_benchmark_your_model --model /path/to/model.tar.gz [--samples 100]

#include "neat.h"

#include <iostream>
#include <stdexcept>
#include <string>

namespace {

bool get_arg(int argc, char** argv, const std::string& key, std::string& out) {
  for (int i = 1; i + 1 < argc; ++i) {
    if (key == argv[i]) {
      out = argv[i + 1];
      return true;
    }
  }
  return false;
}

int parse_int_arg(int argc, char** argv, const std::string& key, int def) {
  std::string value;
  if (!get_arg(argc, argv, key, value))
    return def;
  return std::stoi(value);
}

} // namespace

int main(int argc, char** argv) {
  try {
    std::string model_path;
    if (!get_arg(argc, argv, "--model", model_path)) {
      std::cerr << "Usage: tutorial_003_benchmark_your_model --model <path> [--samples <n>]\n";
      return 1;
    }
    const int samples = parse_int_arg(argc, argv, "--samples", 100);

    // CORE LOGIC
    // STEP load-model
    simaai::neat::Model model(model_path);
    // END STEP

    // STEP run-benchmark
    simaai::neat::BenchmarkReport report = model.benchmark(samples);
    // END STEP

    // STEP read-report
    std::cout << "report_latency_ms=" << report.latency_ms << "\n";
    std::cout << "report_fps=" << report.fps << "\n";
    std::cout << "report_avg_power_watts=" << report.avg_power_watts << "\n";
    std::cout << "report_energy_joules=" << report.energy_joules << "\n";
    // END STEP
    // END CORE LOGIC

    std::cout << "[OK] 003_benchmark_your_model\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}

File: tutorials/003_benchmark_your_model/benchmark_your_model.py
#!/usr/bin/env python3
"""Benchmark a compiled model with deterministic synthetic inputs.

Usage:
  python3 benchmark_your_model.py --model /path/to/model.tar.gz [--samples 100]
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

try:
  import pyneat
except ImportError:
  sys.exit(
      "pyneat is not importable. Either Neat is not installed, or the venv is not activated.\n"
      "Run: source ~/pyneat/bin/activate"
  )


def main(argv: list[str]) -> int:
  ap = argparse.ArgumentParser(description=__doc__)
  ap.add_argument("--model", type=Path, required=True)
  ap.add_argument("--samples", type=int, default=100, help="number of measured synthetic samples")
  args = ap.parse_args(argv[1:])

  # CORE LOGIC
  # STEP load-model
  model = pyneat.Model(str(args.model))
  # END STEP

  # STEP run-benchmark
  report = model.benchmark(args.samples)
  if report.latency_ms <= 0 or report.fps <= 0:
    raise RuntimeError("benchmark produced no measured latency/fps")
  # END STEP

  # STEP read-report
  print(f"report_latency_ms={report.latency_ms}")
  print(f"report_fps={report.fps}")
  print(f"report_avg_power_watts={report.avg_power_watts}")
  print(f"report_energy_joules={report.energy_joules}")
  # END STEP
  # END CORE LOGIC

  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))
